### DISCUSSION TITLE
generate-domains-blocklist.py could be improved to simplify patterns

### DISCUSSION THREAD

**Discussion by 1ennard** [DNSCrypt/dnscrypt-proxy#2402] [Ideas]
The [generate-domains-blocklist.py](https://raw.githubusercontent.com/DNSCrypt/dnscrypt-proxy/master/utils/generate-domains-blocklist/generate-domains-blocklist.py) script is affected by this bug. It not idealy removes overlapping patterns when generating domain blocklists, regarding to the [filtering rules](https://github.com/DNSCrypt/dnscrypt-proxy/wiki/Filters).

## *When* does this occur?
The issue occurs when executing the script to generate a blocklist from a domains-blocklist-local-additions.txt file.

## *Where* does it happen?
Problably in the blocklists_from_config_file function of the generate-domains-blocklist.py script.

## *How* do we replicate the issue?
1. Prepare a domains-blocklist-local-additions.txt file with the following content:
 ~~~
aaa.*
*.aaa
*aaa*
qaaaq.com
aaa.com
aaa.com.com
com.aaa
com.com.aaa

bbb.*
*.bbb
bbb.com
bbb.com.com
com.bbb
com.com.bbb

*ccc*
ccc.com
qcccq.com
~~~

2. Delete all urls from the domains-blocklist.conf
3. Execute the script using:
~~~
python3 generate-domains-blocklist.py -o blocklist.txt
~~~
4. The resulting blocklist.txt will contain:
~~~
########## Blocklist from file:domains-blocklist-local-additions.txt ##########

# Ignored duplicates: 2
# Ignored due to overlapping local patterns: 4

aaa.*
bbb.*
*aaa*
*ccc*
*.aaa
com.aaa
*.bbb
com.bbb
ccc.com
qaaaq.com
qcccq.com
~~~

## Expected behavior (i.e. solution)
The Blocklist should only contain these entries, as they cover all cases:
~~~
*aaa*
*.bbb
bbb.*
*ccc*
~~~

## Other Comments
Wiki page to [Combining Blocklists](https://github.com/DNSCrypt/dnscrypt-proxy/wiki/Combining-Blocklists)

**Comment by jedisct1** [comment]
The tool was only written to remove overlapping names, and, to some extent, names that are matches by existing patterns.

Proving pattern equivalence or that a pattern is a superset of another is not implemented.

Since you appear to be a Python developer, is it a feature you could contribute?

**Reply by flawrist** [reply]
@1ennard, have you managed to enhance `generate-domains-blocklist.py` by yourself?
Please, share.
